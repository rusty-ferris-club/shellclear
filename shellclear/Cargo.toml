[package]
name = "shellclear"
version = "0.4.7"
edition = "2021"
authors = ["Elad-Kaplan <kaplan.elad@gmail.com>"]
readme = "README.md"
repository = "https://github.com/rusty-ferris-club/shellclear"
homepage = "https://github.com/rusty-ferris-club/shellclear"

[dependencies]
clap = { version = "3", features = ["cargo"], optional = true }
serde = "1"
serde_derive = "1"
serde_yaml = "^0.8.24"
serde_regex = "1.1.0"
env_logger = "^0.9.0"
log = "^0.4.17"
anyhow = "1"
console = "^0.15.0"
rayon = "1.5"
regex = "1"
chrono = "0.4"
strum = { version = "0.24", features = ["derive"] }
prettytable-rs = "^0.8"
requestty = "0.4.1"
exitcode = "^1.1.2"
dirs = "4.0"
lazy_static = "1.4.0"
mask-text = "0.1.2"

[dev-dependencies]
insta = { version = "1.17.2", features = ["backtrace", "filters"] }
tempdir = "0.3"

[build-dependencies]
anyhow = "1"

[features]
default = ["cli"]
# list optionals here:
cli = ["clap"]

[[bin]]
name = "shellclear"
required-features = ["cli"]
