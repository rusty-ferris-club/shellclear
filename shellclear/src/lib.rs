extern crate core;

pub use self::{
    emoji::Emojis,
    state::{init, ShellContext},
};

mod emoji;
mod masker;
mod state;

pub mod clearer;
pub mod config;
pub mod data;
pub mod dialog;
pub mod engine;
pub mod exporter;
pub mod shell;
