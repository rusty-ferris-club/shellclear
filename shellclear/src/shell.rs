use std::{fmt, path::Path};

use serde_derive::{Deserialize, Serialize};
use strum::{EnumIter, IntoEnumIterator};

/// List of all supported shells
#[derive(Debug, EnumIter, Clone, Eq, Hash, PartialEq)]
pub enum Shell {
    Bash,
    Zshrc,
    Fish,
    PowerShell,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct FishHistory {
    pub cmd: String,
    pub when: String,
}

impl fmt::Display for Shell {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Self::Bash => write!(f, "bash"),
            Self::Zshrc => write!(f, "zshrc"),
            Self::Fish => write!(f, "fish"),
            Self::PowerShell => write!(f, "powershell"),
        }
    }
}

/// Zsh history file name
const ZSH_HISTORY_FILE_PATH: &str = ".zsh_history";
/// Bash history file name
const BASH_HISTORY_FILE_PATH: &str = ".bash_history";
const FISH_HISTORY_FILE_PATH: &str = ".local/share/fish/fish_history";
const POWERSHELL_HISTORY_FILE_PATH: &str =
    "AppData/Roaming/Microsoft/Windows/PowerShell/PSReadline/ConsoleHost_history.txt";

/// History shell details
#[derive(Clone, Debug)]
pub struct History {
    /// shell type
    pub shell: Shell,
    /// history file path
    pub path: String,
    /// history file name
    pub file_name: String,
}

/// return list of all existing history files
#[must_use]
pub fn get_all_history_files(homedir: &str) -> Vec<History> {
    // return list of existing shells
    Shell::iter()
        .filter_map(|shell| {
            let shell_history_path = get_shell_history_path(&shell, homedir);
            if !Path::new(&shell_history_path).exists() {
                log::debug!("shell {:?} not found", shell);
                return None;
            }
            let file_name = Path::new(&shell_history_path)
                .file_name()?
                .to_str()?
                .to_string();
            Some(History {
                shell,
                path: shell_history_path,
                file_name,
            })
        })
        .collect::<Vec<_>>()
}

// returns all supported shells types
fn get_shell_history_path(shell_type: &Shell, homedir: &str) -> String {
    match shell_type {
        Shell::Bash => Path::new(homedir)
            .join(BASH_HISTORY_FILE_PATH)
            .display()
            .to_string(),
        Shell::Zshrc => Path::new(homedir)
            .join(ZSH_HISTORY_FILE_PATH)
            .display()
            .to_string(),
        Shell::Fish => Path::new(homedir)
            .join(FISH_HISTORY_FILE_PATH)
            .display()
            .to_string(),
        Shell::PowerShell => Path::new(homedir)
            .join(POWERSHELL_HISTORY_FILE_PATH)
            .display()
            .to_string(),
    }
}

#[cfg(test)]
mod state_shell {
    use std::{fs, fs::File};

    use insta::{assert_debug_snapshot, with_settings};
    use tempdir::TempDir;

    use super::{
        get_all_history_files, BASH_HISTORY_FILE_PATH, FISH_HISTORY_FILE_PATH,
        ZSH_HISTORY_FILE_PATH,
    };

    #[test]
    fn can_get_all_history_files() {
        let temp_dir = TempDir::new("terminal").unwrap().path().join("app");

        fs::create_dir_all(&temp_dir).unwrap();
        fs::create_dir_all(temp_dir.join(FISH_HISTORY_FILE_PATH).parent().unwrap()).unwrap();

        File::create(temp_dir.join(ZSH_HISTORY_FILE_PATH)).expect("create failed");
        File::create(temp_dir.join(BASH_HISTORY_FILE_PATH)).expect("create failed");
        File::create(temp_dir.join(FISH_HISTORY_FILE_PATH)).expect("create failed");

        with_settings!({filters => vec![
            (r"//*.+/(app)", "PATH"),
            (r"([C]?:\\.+app\\\\)", "PATH/")// for windows
        ]}, {
            assert_debug_snapshot!(get_all_history_files(&temp_dir.display().to_string()));
        });
    }
}
