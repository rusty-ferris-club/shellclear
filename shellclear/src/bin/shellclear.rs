use std::process::exit;

use anyhow::anyhow;
use console::{style, Style};

use shellclear::{config::Config, dialog, engine, init, Emojis, ShellContext};

mod cmd;

const DEFAULT_ERR_EXIT_CODE: i32 = 1;

pub const BANNER: &str = r#"
     _          _ _      _                 
 ___| |__   ___| | | ___| | ___  __ _ _ __ 
/ __| '_ \ / _ \ | |/ __| |/ _ \/ _` | '__|
\__ \ | | |  __/ | | (__| |  __/ (_| | |   
|___/_| |_|\___|_|_|\___|_|\___|\__,_|_| "#;

fn main() {
    let app = cmd::default::command()
        .subcommand(cmd::config::command())
        .subcommand(cmd::find::command())
        .subcommand(cmd::clear::command())
        .subcommand(cmd::stash::command());

    let matches = app.clone().get_matches();

    let env = env_logger::Env::default().filter_or(
        "LOG",
        matches.value_of("log").unwrap_or(log::Level::Info.as_str()),
    );
    env_logger::init_from_env(env);

    let config = Config::from(matches.value_of("config-dir"));
    // create app config to store state data
    let shells_context = match init() {
        Ok(s) => s,
        Err(e) => {
            log::error!("{}", e);
            exit(1)
        }
    };

    if matches.is_present("init-shell") {
        // In case of an error, we need to suppress the errors to make sure that when
        // new shell is open the users will not get any errors.
        match engine::PatternsEngine::with_config(&config) {
            Ok(engine) => {
                let emojis = Emojis::default();
                if let Ok(shell_commands) =
                    engine.find_history_commands_from_shell_list(&shells_context)
                {
                    let sensitive_commands = shell_commands.get_commands_with_secrets();

                    if sensitive_commands.is_empty() {
                        eprintln!(
                            "{} Your shell is clean from sensitive data!",
                            emojis.confetti
                        );
                    } else {
                        eprintln!(
                            "{} shellclear found {} sensitive commands in your shell history. run \
                             `shellclear find` to see more information",
                            emojis.alarm,
                            style(sensitive_commands.len()).red()
                        );
                    }
                }
            }
            Err(e) => {
                log::debug!("could not init engine config. err: {}", e);
            }
        };
        exit(0)
    }

    if !matches.is_present("no_banner") {
        println!(
            "{}{}",
            style(BANNER).magenta(),
            style(app.get_version().unwrap_or("")).dim()
        );
    }

    let res = match matches.subcommand() {
        None => Err(anyhow!("command not found")),
        Some(tup) => match tup {
            ("config", subcommand_matches) => cmd::config::run(subcommand_matches, &config),
            ("find", subcommand_matches) => {
                cmd::find::run(subcommand_matches, &shells_context, &config)
            }
            ("clear", subcommand_matches) => {
                cmd::clear::run(subcommand_matches, &shells_context, &config)
            }
            ("stash", subcommand_matches) => {
                cmd::stash::run(subcommand_matches, select_shell(&shells_context))
            }
            _ => unreachable!(),
        },
    };

    let exit_with = match res {
        Ok(cmd) => {
            if let Some(message) = cmd.message {
                let style = if exitcode::is_success(cmd.code) {
                    Style::new().green()
                } else {
                    Style::new().red()
                };
                eprintln!("\r\n{}\r\n", style.apply_to(message));
            }
            cmd.code
        }
        Err(e) => {
            log::debug!("{:?}", e);
            DEFAULT_ERR_EXIT_CODE
        }
    };
    exit(exit_with)
}

fn select_shell(shell_contexts: &Vec<ShellContext>) -> &ShellContext {
    if shell_contexts.len() == 1 {
        return &shell_contexts[0];
    }
    let selections = shell_contexts
        .iter()
        .map(|f| format!("{:?}     : {}", f.history.shell, f.history.path))
        .collect::<Vec<_>>();

    match dialog::select("Pick your shell", &selections) {
        Ok(selection) => &shell_contexts[selection],
        Err(e) => {
            log::debug!("prompter select err: {:?} ", e);
            exit(1)
        }
    }
}
