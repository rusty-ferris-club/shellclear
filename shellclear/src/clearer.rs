use std::fmt::Write;
use std::fs::write;
use std::time::Instant;

use anyhow::Result;

use crate::engine::ShellCommands;
use crate::ShellContext;

pub struct Clearer {}

impl Clearer {
    /// Load sensitive ignores file
    ///
    /// # Errors
    ///
    /// Will return `Err` when the history file cannot be opened / written to
    pub fn write_findings(
        shells_context: &[ShellContext],
        commands: &ShellCommands,
        remove: bool,
    ) -> Result<()> {
        for context in shells_context {
            let start = Instant::now();
            let mut cleared_history: String = String::new();

            for command in commands
                .get_commands_per_shell(&context.history.shell)
                .unwrap_or(&vec![])
            {
                if remove {
                    if command.detections.is_empty() {
                        writeln!(&mut cleared_history, "{}", command.data)?;
                    }
                } else {
                    writeln!(&mut cleared_history, "{}", command.data)?;
                }
            }

            if !cleared_history.is_empty() {
                write(&context.history.path, cleared_history)?;
                log::debug!(
                    "time elapsed for backup existing file and write a new history to shell : {:?}",
                    start.elapsed()
                );
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::Path;

    use insta::assert_debug_snapshot;
    use regex::Regex;
    use tempdir::TempDir;

    use crate::data::{Command, Detection};
    use crate::shell::History;
    use crate::shell::Shell::Zshrc;

    use super::*;

    fn mock_state(dir: &Path) -> (ShellCommands, Vec<ShellContext>) {
        let state_context = vec![ShellContext {
            app_folder_path: "mock".to_string(),
            history: History {
                shell: Zshrc,
                file_name: "mock".to_string(),
                path: dir.to_str().unwrap().into(),
            },
        }];

        // We test one finding with a secret and one without a secret
        // The secret should be filtered out when using remove = true param
        let findings = vec![
            Command {
                command: "mock".to_string(),
                data: "mock".to_string(),
                detections: vec![],
                shell_type: Zshrc,
                secrets: vec![],
            },
            Command {
                command: "should be removed".to_string(),
                data: "should be removed".to_string(),
                detections: vec![Detection {
                    name: "mock".to_string(),
                    secret_group: 0,
                    id: "mock-id".to_string(),
                    test: Regex::new("").unwrap(),
                }],
                shell_type: Zshrc,
                secrets: vec!["mock".to_string()],
            },
        ];

        let mut commands = ShellCommands::default();

        commands.add_commands(&Zshrc, findings);

        (commands, state_context)
    }

    #[test]
    fn remove_sensitive_commands() {
        let dir = TempDir::new("clearer").unwrap();
        let history = dir.path().join("history");

        let (commands, state_context) = mock_state(&history);

        Clearer::write_findings(&state_context, &commands, true).unwrap();

        let content = fs::read_to_string(history).unwrap();

        assert_debug_snapshot!(content);

        dir.close().unwrap();
    }

    #[test]
    fn persist_sensitive_commands() {
        let dir = TempDir::new("clearer").unwrap();
        let history = dir.path().join("history");

        let (commands, state_context) = mock_state(&history);

        Clearer::write_findings(&state_context, &commands, false).unwrap();

        let content = fs::read_to_string(history).unwrap();

        assert_debug_snapshot!(content);

        dir.close().unwrap();
    }
}
