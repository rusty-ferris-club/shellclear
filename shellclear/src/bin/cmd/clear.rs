use anyhow::Result;
use clap::{Arg, ArgMatches, Command};

use shellclear::clearer::Clearer;
use shellclear::{config::Config, engine, Emojis, ShellContext};

pub fn command() -> Command<'static> {
    Command::new("clear")
        .about("Remove or mask the findings from shell history")
        .arg(
            Arg::new("backup")
                .long("backup")
                .help("Backup history file before delete commands")
                .takes_value(false),
        )
        .arg(
            Arg::new("remove")
                .long("remove")
                .help("Remove history that contains secrets")
                .takes_value(false),
        )
}

pub fn run(
    matches: &ArgMatches,
    shells_context: &Vec<ShellContext>,
    config: &Config,
) -> Result<shellclear::data::CmdExit> {
    let en = engine::PatternsEngine::with_config(config)?;

    for shell_context in shells_context {
        if matches.is_present("backup") {
            match shell_context.backup() {
                Ok(path) => log::debug!("history backup successful: {}", path),
                Err(e) => {
                    return Ok(shellclear::data::CmdExit {
                        code: 1,
                        message: Some(format!(
                            "could not backup shell {:?} history. err: {:?}",
                            shell_context.history.shell, e
                        )),
                    });
                }
            }
        }
    }

    let (findings, sensitive_commands) =
        en.find_history_commands_from_shell_list(shells_context)?;
    let emojis = Emojis::default();

    Clearer::write_findings(shells_context, &findings, matches.is_present("remove"))?;

    if sensitive_commands.is_empty() {
        return Ok(shellclear::data::CmdExit {
            code: exitcode::OK,
            message: Some(format!(
                "{} Your shell is clean from sensitive data!",
                emojis.confetti
            )),
        });
    };

    let message = format!(
        " {} shellclear cleared {} sensitive commands",
        emojis.alarm,
        sensitive_commands.len()
    );

    Ok(shellclear::data::CmdExit {
        code: exitcode::OK,
        message: Some(message),
    })
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::fs::File;
    use std::io::Write;

    use insta::assert_debug_snapshot;
    use tempdir::TempDir;

    use shellclear::shell::{History, Shell};

    use crate::cmd::clear::run;

    use super::*;

    const TEST_SENSITIVE_COMMANDS: &str = r###"
- name: Find me
  secret_group: 0
  test: FIND_ME=
    "###;

    const TEMP_HISTORY_LINES_CONTENT: &str = "history
ls
echo 'hello you'
rm -f ./file.txt
export FIND_ME=token
";

    const TEMP_HISTORY_FISH: &str = r#"---
- cmd: history
  when: "1656438759"
- cmd: ls
  when: "1656438760"
- cmd: echo 'hello you'
  when: "1656438760"
- cmd: rm -f ./file.txt
  when: "1656438760"
- cmd: export FIND_ME=token
  when: "1656438760"
"#;

    fn create_mock_state(
        temp_dir: &TempDir,
        content: &str,
        rules: &str,
        shell_type: Shell,
    ) -> ShellContext {
        let app_folder = temp_dir.path().join("shellclear");
        let history_file_name = "history";
        let history_file_path = app_folder.join(history_file_name);
        fs::create_dir_all(&app_folder).unwrap();

        let mut f = File::create(&history_file_path).unwrap();
        f.write_all(content.as_bytes()).unwrap();
        f.sync_all().unwrap();

        let sensitive_patterns_path = app_folder.join("sensitive-patterns.yaml");

        let mut f = File::create(sensitive_patterns_path).unwrap();
        f.write_all(rules.as_bytes()).unwrap();
        f.sync_all().unwrap();

        ShellContext {
            app_folder_path: app_folder.display().to_string(),
            history: History {
                shell: shell_type,
                path: history_file_path.display().to_string(),
                file_name: history_file_name.to_string(),
            },
        }
    }

    #[test]
    fn can_clear_command_by_lines() {
        let temp_dir = TempDir::new("clear").unwrap();

        let state_context = create_mock_state(
            &temp_dir,
            TEMP_HISTORY_LINES_CONTENT,
            TEST_SENSITIVE_COMMANDS,
            Shell::Bash,
        );

        let args = command().get_matches_from(vec!["clear", "--remove"]);

        let result = run(
            &args,
            &vec![state_context.clone()],
            &Config::from(temp_dir.path().to_str()),
        )
        .unwrap()
        .message
        .unwrap()
        .replace(|c: char| !c.is_ascii(), "");

        assert_debug_snapshot!(result);
        assert_debug_snapshot!(fs::read_to_string(state_context.history.path).unwrap());
        temp_dir.close().unwrap();
    }

    #[test]
    fn can_clear_find_fish() {
        let temp_dir = TempDir::new("clear").unwrap();

        let state_context = create_mock_state(
            &temp_dir,
            TEMP_HISTORY_FISH,
            TEST_SENSITIVE_COMMANDS,
            Shell::Fish,
        );

        let args = command().get_matches_from(vec!["clear", "--remove"]);

        let result = run(
            &args,
            &vec![state_context.clone()],
            &Config::from(temp_dir.path().to_str()),
        )
        .unwrap()
        .message
        .unwrap()
        .replace(|c: char| !c.is_ascii(), "");

        assert_debug_snapshot!(result);
        assert_debug_snapshot!(fs::read_to_string(state_context.history.path).unwrap());
        temp_dir.close().unwrap();
    }
}
