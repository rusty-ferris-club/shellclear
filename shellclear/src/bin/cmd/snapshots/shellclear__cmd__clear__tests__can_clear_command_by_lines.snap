---
source: shellclear/src/bin/cmd/clear.rs
assertion_line: 185
expression: result
---
Ok(
    CmdExit {
        code: 0,
        message: Some(
            " ðŸš¨ shellclear cleared 1 sensitive commands",
        ),
    },
)
